src/amdgpu_libdrm.cpp

```diff
@@ -0,0 +1,112 @@
+#include <deque>
+#include <mutex>
+#include <thread>
+#include <chrono>
+#include <xf86drm.h>
+#include <libdrm/amdgpu.h>
+#include <spdlog/spdlog.h>
+#include <fcntl.h>
+#include "gpu.h"
+
+#include "amdgpu_libdrm.h"
+
+std::string dri_device_path;
+bool do_libdrm_sampling = false;
+
+std::deque<struct libdrm_sample> sample_buf(LIBDRM_SAMPLE_BUF_SIZE, {0});
+std::mutex sample_buf_m;
+
+amdgpu_device_handle amdgpu_handle;
+
+static void libdrm_do_sample(libdrm_sample *sample) {
+    uint32_t registers;
+    amdgpu_read_mm_registers(amdgpu_handle, LIBDRM_GRBM_STATUS / 4, 1, 0xffffffff, 0, &registers);
+
+    if (registers & LIBDRM_GRBM_BUSY_BIT) sample->busy_bit = true;
+}
+
+static void libdrm_thread() {
+    while (true) {
+        auto start_time = std::chrono::system_clock::now().time_since_epoch();
+
+        struct libdrm_sample sample {0};
+        libdrm_do_sample(&sample);
+        
+        sample_buf_m.lock();
+        sample_buf.pop_front();
+        sample_buf.push_back(sample);
+        sample_buf_m.unlock();
+
+        auto end_time = std::chrono::system_clock::now().time_since_epoch();
+        auto sleep_duration = std::chrono::microseconds(LIBDRM_SAMPLE_DELAY) - (end_time - start_time);
+        if (sleep_duration > std::chrono::nanoseconds(0)) {
+            std::this_thread::sleep_for(sleep_duration);
+        }
+    }
+}
+
+static int libdrm_initialize() {
+    drmDevicePtr devices[LIBDRM_MAX_DEVICES];
+    int device_count = drmGetDevices2(0, devices, LIBDRM_MAX_DEVICES);
+    if (device_count < 0) {
+        SPDLOG_ERROR("drmGetDevices2 failed");       
+        return -1;
+    }
+
+    char *renderd_node = nullptr;
+    for (int i = 0; i < device_count; i++) {
+        if ((devices[i]->available_nodes & 0b101) != 0b101) {
+            continue;
+        }
+
+        if (devices[i]->nodes[DRM_NODE_PRIMARY] == dri_device_path) {
```

Comment:
Disclaimer C++ is rusty: String comparison usually happens with `strcmp()`. This will compare the pointers.

Reply:
`dri_device_path` is an std::string and not a char*. Would that still compare just the pointers?

Reply:
Right, didn't see that. AFAICT there should be a `bool operator== (const std::string& c1, const char *)` in the system headers, making this work correctly. Ignore my comment.